File: LeetCode/0283-move-zeroes/0283-move-zeroes.cpp
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int N = nums.size();
        if (N == 1)
            return;
        int p1 = 0, p2 = 1;
        while (p1 < N && p2 < N) {
            if (nums[p1] == 0) {
                if (p2 <= p1)
                    p2 = p1 + 1;
                else if (nums[p2] != 0)
                    swap(nums[p1], nums[p2]);
            }
            if (nums[p1] != 0)
                p1++;
            if (nums[p2] == 0)
                p2++;
        }
    }
};

File: LeetCode/0283-move-zeroes/0283-move-zeroes.py
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        left = 0
        right = 1
        n = len(nums)

        while left < n and right < n:
            while left < n and nums[left] != 0:
                left += 1
            if left >= right:
                right = left + 1
            while right < n and nums[right] == 0:
                right += 1
            
            if left >= n or right >= n: 
                break

            nums[left], nums[right] = nums[right], nums[left]
            left += 1
            right += 1
        
        return nums
        